### PR TITLE
Deprecate Redundant property enabled_focus_mode

### LINKED ISSUE #41529: BaseButton property "enabled_focus_mode" is redundant
**Godot version:**

3.2

**OS/device including version:**

no specific

**Issue description:**

#4300 added a prop to BaseButton called "enabled_focus_mode" for setting the focus mode for enabled buttons. At that time, buttons would set focus mode to FOCUS_NONE on disabling and set to "enabled_focus_mode" when enabling. However, #16152 removed above logic, but leaving the prop "enabled_focus_mode" there.

Currently, this prop is really confusing that it would not have any effect when calling "BaseButton::set_enabled_focus_mode" on a disabled button without any warning, and has the exactly same behavior as the "Control::set_focus_mode" on an enabled button.

### PR DESCRIPTION
**Pull request by FIF15**
of BaseButton
see #41529 for details
this closes #41529, related to #41576

### CODE REVIEW

File: scene/gui/base_button.cpp [hunk 1]
```diff
@@ -323,9 +323,7 @@ int BaseButton::get_button_mask() const {
 void BaseButton::set_enabled_focus_mode(FocusMode p_mode) {
 
 	enabled_focus_mode = p_mode;
-	if (!status.disabled) {
-		set_focus_mode(p_mode);
-	}
+	WARN_DEPRECATED_MSG("BaseButton's Enabled Focus Mode property has been deprecated due to redundancy and will be removed in Godot 4.0.");
```
Comment (akien-mga) [inline scene/gui/base_button.cpp:326]: Maybe refer to `Control.set_focus_mode` which should be used instead?

File: scene/gui/link_button.cpp [hunk 2]
```diff
@@ -140,6 +140,6 @@ void LinkButton::_bind_methods() {
 
 LinkButton::LinkButton() {
 	underline_mode = UNDERLINE_MODE_ALWAYS;
-	set_enabled_focus_mode(FOCUS_NONE);
+	set_focus_mode(FOCUS_NONE);
```
Comment (akien-mga) [inline scene/gui/link_button.cpp:143]: Same comment as for the `master` PR, should not be necessary as it's the default.

File: doc/classes/BaseButton.xml [hunk 3]
```diff
@@ -52,7 +52,7 @@
 			If [code]true[/code], the button is in disabled state and can't be clicked or toggled.
 		</member>
 		<member name="enabled_focus_mode" type="int" setter="set_enabled_focus_mode" getter="get_enabled_focus_mode" enum="Control.FocusMode" default="2">
-			Focus access mode to use when switching between enabled/disabled (see [member Control.focus_mode] and [member disabled]).
+			[i]Deprecated.[/i] This property has been deprecated due to redundancy and no longer has any effect when set.
```
Comment (akien-mga) [inline doc/classes/BaseButton.xml:55]: It would be good to tell users what to use instead.

**Comment by FIF15** [comment]
@akien-mga, your suggestions are applied. Thanks.

File: doc/classes/BaseButton.xml [hunk 4]
```diff
@@ -52,7 +52,7 @@
 			If [code]true[/code], the button is in disabled state and can't be clicked or toggled.
 		</member>
 		<member name="enabled_focus_mode" type="int" setter="set_enabled_focus_mode" getter="get_enabled_focus_mode" enum="Control.FocusMode" default="2">
-			Focus access mode to use when switching between enabled/disabled (see [member Control.focus_mode] and [member disabled]).
+			[i]Deprecated.[/i] This property has been deprecated due to redundancy and no longer has any effect when set. Please use [code]Control.set_focus_mode[/code] instead.
```
Comment (akien-mga) [inline doc/classes/BaseButton.xml:55]:
```suggestion
			[i]Deprecated.[/i] This property has been deprecated due to redundancy and no longer has any effect when set. Please use [member Control.focus_mode] instead.
```

**Comment by akien-mga** [comment]
Could you squash the commits into one?

**Comment by FIF15** [comment]
> Could you squash the commits into one?

sure

File: scene/gui/base_button.cpp [hunk 5]
```diff
@@ -323,9 +323,7 @@ int BaseButton::get_button_mask() const {
 void BaseButton::set_enabled_focus_mode(FocusMode p_mode) {
 
 	enabled_focus_mode = p_mode;
-	if (!status.disabled) {
-		set_focus_mode(p_mode);
-	}
```
Comment (akien-mga) [inline scene/gui/base_button.cpp:328]: Now that I think about it, shouldn't this be kept for compatibility? The warning is good to tell users that they should change their code, but the code should still work properly without change in the `3.2` branch.

Comment (FIF15) [reply scene/gui/base_button.cpp:328]: Yes, we can keep that. Does the warning message need any change?

Comment (akien-mga) [reply scene/gui/base_button.cpp:328]: No I think the message is good. It tells users that they should change their code, but doesn't imply that the existing code is invalidated.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**